File: src/FluentCards/InputToggleBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating InputToggle elements.
/// </summary>
public class InputToggleBuilder
{
    private readonly InputToggle _input = new();

    /// <summary>
    /// Sets the unique identifier for the input.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithId(string id)
    {
        _input.Id = id;
        return this;
    }

    /// <summary>
    /// Sets the label for the input.
    /// </summary>
    /// <param name="label">The input label.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithLabel(string label)
    {
        _input.Label = label;
        return this;
    }

    /// <summary>
    /// Sets the title displayed next to the toggle.
    /// </summary>
    /// <param name="title">The title text.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithTitle(string title)
    {
        _input.Title = title;
        return this;
    }

    /// <summary>
    /// Sets the value to submit when the toggle is on.
    /// </summary>
    /// <param name="valueOn">The value when on.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithValueOn(string valueOn)
    {
        _input.ValueOn = valueOn;
        return this;
    }

    /// <summary>
    /// Sets the value to submit when the toggle is off.
    /// </summary>
    /// <param name="valueOff">The value when off.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithValueOff(string valueOff)
    {
        _input.ValueOff = valueOff;
        return this;
    }

    /// <summary>
    /// Sets the default value for the toggle.
    /// </summary>
    /// <param name="value">The default value.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithValue(string value)
    {
        _input.Value = value;
        return this;
    }

    /// <summary>
    /// Sets whether the input is required.
    /// </summary>
    /// <param name="required">True if required, false otherwise.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder IsRequired(bool required = true)
    {
        _input.IsRequired = required;
        return this;
    }

    /// <summary>
    /// Sets the error message to display when validation fails.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithErrorMessage(string message)
    {
        _input.ErrorMessage = message;
        return this;
    }

    /// <summary>
    /// Sets whether the title text should wrap.
    /// </summary>
    /// <param name="wrap">True to wrap, false otherwise.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputToggleBuilder WithWrap(bool wrap)
    {
        _input.Wrap = wrap;
        return this;
    }

    /// <summary>
    /// Builds and returns the configured InputToggle.
    /// </summary>
    /// <returns>The configured InputToggle instance.</returns>
    public InputToggle Build()
    {
        return _input;
    }
}

File: go/fluentcards/input_toggle_builder.go
package fluentcards

// InputToggleBuilder builds an Input.Toggle Adaptive Card element.
type InputToggleBuilder struct {
	data map[string]any
}

func newInputToggleBuilder() *InputToggleBuilder {
	return &InputToggleBuilder{data: map[string]any{"type": "Input.Toggle", "id": "", "title": ""}}
}

func (b *InputToggleBuilder) WithID(id string) *InputToggleBuilder {
	b.data["id"] = id
	return b
}

func (b *InputToggleBuilder) WithTitle(title string) *InputToggleBuilder {
	b.data["title"] = title
	return b
}

func (b *InputToggleBuilder) WithLabel(label string) *InputToggleBuilder {
	b.data["label"] = label
	return b
}

func (b *InputToggleBuilder) WithValue(value string) *InputToggleBuilder {
	b.data["value"] = value
	return b
}

func (b *InputToggleBuilder) WithValueOn(valueOn string) *InputToggleBuilder {
	b.data["valueOn"] = valueOn
	return b
}

func (b *InputToggleBuilder) WithValueOff(valueOff string) *InputToggleBuilder {
	b.data["valueOff"] = valueOff
	return b
}

func (b *InputToggleBuilder) WithWrap(wrap bool) *InputToggleBuilder {
	b.data["wrap"] = wrap
	return b
}

func (b *InputToggleBuilder) WithIsRequired(isRequired bool) *InputToggleBuilder {
	b.data["isRequired"] = isRequired
	return b
}

func (b *InputToggleBuilder) WithErrorMessage(errorMessage string) *InputToggleBuilder {
	b.data["errorMessage"] = errorMessage
	return b
}

func (b *InputToggleBuilder) WithSpacing(spacing Spacing) *InputToggleBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

func (b *InputToggleBuilder) WithIsVisible(isVisible bool) *InputToggleBuilder {
	b.data["isVisible"] = isVisible
	return b
}

func (b *InputToggleBuilder) WithSeparator(separator bool) *InputToggleBuilder {
	b.data["separator"] = separator
	return b
}

func (b *InputToggleBuilder) WithHeight(height string) *InputToggleBuilder {
	b.data["height"] = height
	return b
}

func (b *InputToggleBuilder) WithFallback(fallback any) *InputToggleBuilder {
	b.data["fallback"] = fallback
	return b
}

func (b *InputToggleBuilder) WithRequires(key, version string) *InputToggleBuilder {
	reqs, ok := b.data["requires"].(map[string]any)
	if !ok {
		reqs = map[string]any{}
	}
	reqs[key] = version
	b.data["requires"] = reqs
	return b
}

func (b *InputToggleBuilder) WithRtl(rtl bool) *InputToggleBuilder {
	b.data["rtl"] = rtl
	return b
}

func (b *InputToggleBuilder) WithLabelPosition(position InputLabelPosition) *InputToggleBuilder {
	b.data["labelPosition"] = string(position)
	return b
}

func (b *InputToggleBuilder) WithLabelWidth(width string) *InputToggleBuilder {
	b.data["labelWidth"] = width
	return b
}

func (b *InputToggleBuilder) WithInputStyle(style InputStyle) *InputToggleBuilder {
	b.data["inputStyle"] = string(style)
	return b
}

func (b *InputToggleBuilder) Build()Card {
	return b.data
}
